infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/billing/AzureBillingResourceUsage.java

```diff
@@ -0,0 +1,281 @@
+        String lastResourceUsageHistory = null;
+        // while startDateTime < endDateTime
+        while (startDateTime.isBefore(endDateTime)) {
+
+            // Query the resource usage history over the current period
+            String startDateTimeStr = formatter.format(startDateTime);
+            String endDateTimeStr = formatter.format(endDateTime);
+            lastResourceUsageHistory = queryResourceUsageHistory(startDateTimeStr,
+                                                                 endDateTimeStr,
+                                                                 azureBillingCredentials.renewOrOnlyGetAccessToken(false));
+            LOGGER.debug("AzureBillingResourceUsage getLastResourceUsageHistory considering [" + startDateTimeStr +
+                         ";" + endDateTimeStr + "] = " + lastResourceUsageHistory);
+
+            JsonObject jsonObject = JSON_PARSER.parse(lastResourceUsageHistory).getAsJsonObject();
+
+            // HISTORY RETRIEVED !!
+            if (JSON_PARSER.parse(lastResourceUsageHistory).getAsJsonObject().has("value")) {
+                this.resourceUsageReportedEndDateTime = endDateTime;
+                LOGGER.debug("AzureBillingResourceUsage getLastResourceUsageHistory resource usage history is finally retrieved!");
+                return lastResourceUsageHistory;
+
+            } else if (jsonObject.has("error")) { // HISTORY NOT RETRIEVED BUT TRY AGAIN !!
+
+                String queryErrorCodeMessage = jsonObject.get("error").getAsJsonObject().get("code").getAsString();
+
+                if (queryErrorCodeMessage.equals("ExpiredAuthenticationToken")) {
+                    LOGGER.debug("AzureBillingResourceUsage getLastResourceUsageHistory ExpiredAuthenticationToken");
+                    azureBillingCredentials.renewOrOnlyGetAccessToken(true);
+                    continue;
+                } else if (queryErrorCodeMessage.equals("ProcessingNotCompleted")) {
+                    endDateTime = endDateTime.minusHours(1);
+                    LOGGER.debug("AzureBillingResourceUsage getLastResourceUsageHistory ProcessingNotCompleted new endDateTime " +
+                                 endDateTime);
+                    continue;
+                }
+            }
+
+            // HISTORY WILL NEVER BE RETRIEVED, throw an Exception to stop the periodical getter threads !!
+            LOGGER.error("AzureBillingResourceUsage getLastResourceUsageHistory AzureBillingException " +
+                         lastResourceUsageHistory);
+            throw new AzureBillingException(lastResourceUsageHistory);
+        }
+        // DID NOT FIND AN ENDDATETIME TO GET THE HISTORY.. WILL TRY AT THE NEW PERIODICAL CALL !
+        LOGGER.debug("AzureBillingResourceUsage getLastResourceUsageHistory cannot find an history for this period since " +
+                     formatter.format(startDateTime) + " >= " + formatter.format(endDateTime));
+        return null;
```

Comment:
in the else block you throw an exception
why do you return null ?

Reply:
Because the method must return a String.